src/sorted_set.h

```diff
@@ -0,0 +1,94 @@
+#ifndef SORTED_SET_H
+#define SORTED_SET_H
+
+#include <algorithm>
+#include <cassert>
+#include <vector>
+
+/// Modifying a key in associative containers leads to UB. This is an std::set
+/// reimplementation with worse characteristics (such as re-sorting the
+/// container on every push), but one that allows for elements to be
+/// out-of-order (which can happen after a key was modified)
+template<class Elem, class Compare>
+struct SortedSet {
+  SortedSet(): comp(Compare()) {}
+  size_t size() const {
+    return data.size();
+  }
+
+  bool has_element(const Elem& elem) const {
+    return std::find(data.begin(), data.end(), elem) != data.end();
+  }
+
+  void insert(const Elem& elem) {
+    if (has_element(elem))
+      return;
+    data.push_back(elem);
+    sort(data.begin(), data.end(), comp);
```

Comment:
This is suboptimal. `sort` may require more than O(n) operations here. Something like the following is guaranteed to take O(n) operations:
```suggestion
    std::vector<Elem>::iterator pos = std::lower_bound(data.begin(), data.end(), elem, comp);
    data.insert(pos, elem);
```
You could alse merge it with searching if element already exists, to avoid double searching.

Reply:
By the way, `flat_map` and `flat_set` (sorted containers) are scheduled to be included in C++20. Proposal is here https://wg21.link/p1222 See also how reference implementation does `insert` here: https://github.com/WG21-SG14/SG14/blob/d12d8acc59071cfdc44bc5d6ad66d43e1db99d0a/SG14/flat_set.h#L394-L402